src/cargo/util/graph.rs

```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
     }
 }
 
-impl<N: Eq + Ord + Clone, E: Eq + Clone> PartialEq for Graph<N, E> {
+impl<N: Eq + Hash + Clone, E: Eq + Clone + PartialEq> PartialEq for Graph<N, E> {
     fn eq(&self, other: &Graph<N, E>) -> bool {
         self.nodes.eq(&other.nodes)
     }
 }
-impl<N: Eq + Ord + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
+impl<N: Eq + Hash + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
 
-impl<N: Eq + Ord + Clone, E: Clone> Clone for Graph<N, E> {
-    fn clone(&self) -> Graph<N, E> {
-        Graph {
-            nodes: self.nodes.clone(),
+/// This is a directed Graph structure, that builds on the `Graph`'s "append only" internals
+/// to provide:
+///  - `O(ln(number of clones))` clone, no dependence on the size of the graph
+///  - the clone has no overhead to read the `Graph` as it was
+///  - no overhead over using the `Graph` directly when modifying the biggest clone
+/// Is this too good to be true?
+///  - If a modification (`&mut` method) is done to a smaller older clone then a full `O(N)`
+///    deep clone will happen internally.
+pub struct StackGraph<N: Clone, E: Clone> {
+    /// The `Graph` shared by all clones, this `StackGraph` refers to only the prefix of size `age`
+    inner: Rc<RefCell<Graph<N, E>>>,
+    /// The shared list of all extant references to the same `Graph`.
+    /// The largest one is allowed to reset and expend the `Graph` without doing a deep clone.
+    /// There can be more then one clone at the same age, so each age is associated with a count.
+    other_refs: Rc<RefCell<BTreeMap<GraphAge, usize>>>,
```

Comment:
I can't really explain why, but this seems like a sub-optimal way to track if we're the largest graph. I'd expect you could have something less complex or more intrinsic?